<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>SAM and BAM · BioAlignments.jl</title><link href="https://cdnjs.cloudflare.com/ajax/libs/normalize/4.2.0/normalize.min.css" rel="stylesheet" type="text/css"/><link href="https://fonts.googleapis.com/css?family=Lato|Roboto+Mono" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/styles/default.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL="."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.2.0/require.min.js" data-main="assets/documenter.js"></script><script src="siteinfo.js"></script><script src="../versions.js"></script><link href="assets/documenter.css" rel="stylesheet" type="text/css"/></head><body><nav class="toc"><a href="index.html"><img class="logo" src="assets/logo.png" alt="BioAlignments.jl logo"/></a><h1>BioAlignments.jl</h1><select id="version-selector" onChange="window.location.href=this.value" style="visibility: hidden"></select><form class="search" id="search-form" action="search.html"><input id="search-query" name="q" type="text" placeholder="Search docs"/></form><ul><li><a class="toctext" href="index.html">Home</a></li><li><a class="toctext" href="alignments.html">Alignment representation</a></li><li><a class="toctext" href="pairalign.html">Pairwise alignment</a></li><li><span class="toctext">IO</span><ul><li class="current"><a class="toctext" href="hts-files.html">SAM and BAM</a><ul class="internal"><li><a class="toctext" href="#Introduction-1">Introduction</a></li><li><a class="toctext" href="#Reading-SAM-and-BAM-files-1">Reading SAM and BAM files</a></li><li><a class="toctext" href="#SAM-and-BAM-Headers-1">SAM and BAM Headers</a></li><li><a class="toctext" href="#SAM-and-BAM-Records-1">SAM and BAM Records</a></li><li><a class="toctext" href="#Accessing-auxiliary-data-1">Accessing auxiliary data</a></li><li><a class="toctext" href="#Getting-records-in-a-range-1">Getting records in a range</a></li><li><a class="toctext" href="#Getting-records-overlapping-genomic-features-1">Getting records overlapping genomic features</a></li><li><a class="toctext" href="#Writing-files-1">Writing files</a></li></ul></li></ul></li><li><a class="toctext" href="references.html">API Reference</a></li></ul></nav><article id="docs"><header><nav><ul><li>IO</li><li><a href="hts-files.html">SAM and BAM</a></li></ul><a class="edit-page" href="https://github.com/BioJulia/BioAlignments.jl/blob/master/docs/src/hts-files.md"><span class="fa"></span> Edit on GitHub</a></nav><hr/><div id="topbar"><span>SAM and BAM</span><a class="fa fa-bars" href="#"></a></div></header><h1><a class="nav-anchor" id="SAM-and-BAM-1" href="#SAM-and-BAM-1">SAM and BAM</a></h1><h2><a class="nav-anchor" id="Introduction-1" href="#Introduction-1">Introduction</a></h2><p>High-throughput sequencing (HTS) technologies generate a large amount of data in the form of a large number of nucleotide sequencing reads. One of the most common tasks in bioinformatics is to align these reads against known reference genomes, chromosomes, or contigs. BioAlignments provides several data formats commonly used for this kind of task.</p><p>BioAlignments offers high-performance tools for SAM and BAM file formats, which are the most popular file formats.</p><p>If you have questions about the SAM and BAM formats or any of the terminology used when discussing these formats, see the published [specification][samtools-spec], which is maintained by the [samtools group][samtools].</p><p>A very very simple SAM file looks like the following:</p><pre><code class="language-none">@HD VN:1.6 SO:coordinate
@SQ SN:ref LN:45
r001   99 ref  7 30 8M2I4M1D3M = 37  39 TTAGATAAAGGATACTG *
r002    0 ref  9 30 3S6M1P1I4M *  0   0 AAAAGATAAGGATA    *
r003    0 ref  9 30 5S6M       *  0   0 GCCTAAGCTAA       * SA:Z:ref,29,-,6H5M,17,0;
r004    0 ref 16 30 6M14N5M    *  0   0 ATAGCTTCAGC       *
r003 2064 ref 29 17 6H5M       *  0   0 TAGGC             * SA:Z:ref,9,+,5S6M,30,1;
r001  147 ref 37 30 9M         =  7 -39 CAGCGGCAT         * NM:i:1</code></pre><p>Where the first two lines are part of the &quot;header&quot;, and the following lines are &quot;records&quot;. Each record describes how a read aligns to some reference sequence. Sometimes one record describes one read, but there are other cases like chimeric reads and split alignments, where multiple records apply to one read. In the example above, <code>r003</code> is a chimeric read, and <code>r004</code> is a split alignment, and <code>r001</code> are mate pair reads. Again, we refer you to the official [specification][samtools-spec] for more details.</p><p>A BAM file stores this same information but in a binary and compressible format that does not make for pretty printing here!</p><h2><a class="nav-anchor" id="Reading-SAM-and-BAM-files-1" href="#Reading-SAM-and-BAM-files-1">Reading SAM and BAM files</a></h2><p>A typical script iterating over all records in a file looks like below:</p><pre><code class="language-julia">using BioAlignments

# Open a BAM file.
reader = open(BAM.Reader, &quot;data.bam&quot;)

# Iterate over BAM records.
for record in reader
    # `record` is a BAM.Record object.
    if BAM.ismapped(record)
        # Print the mapped position.
        println(BAM.refname(record), &#39;:&#39;, BAM.position(record))
    end
end

# Close the BAM file.
close(reader)</code></pre><p>The size of a BAM file is often extremely large. The iterator interface demonstrated above allocates an object for each record and that may be a bottleneck of reading data from a BAM file. In-place reading reuses a pre-allocated object for every record and less memory allocation happens in reading:</p><pre><code class="language-julia">reader = open(BAM.Reader, &quot;data.bam&quot;)
record = BAM.Record()
while !eof(reader)
    read!(reader, record)
    # do something
end</code></pre><h2><a class="nav-anchor" id="SAM-and-BAM-Headers-1" href="#SAM-and-BAM-Headers-1">SAM and BAM Headers</a></h2><p>Both <code>SAM.Reader</code> and <code>BAM.Reader</code> implement the <code>header</code> function, which returns a <code>SAM.Header</code> object. To extract certain information out of the headers, you can use the <code>find</code> method on the header to extract information according to SAM/BAM tag. Again we refer you to the [specification][samtools-spec] for full details of all the different tags that can occur in headers, and what they mean.</p><p>Below is an example of extracting all the info about the reference sequences from the BAM header. In SAM/BAM, any description of a reference sequence is stored in the header, under a tag denoted <code>SQ</code> (think <code>reference SeQuence</code>!).</p><pre><code class="language-jlcon">julia&gt; reader = open(SAM.Reader, &quot;data.sam&quot;);

julia&gt; find(header(reader), &quot;SQ&quot;)
7-element Array{Bio.Align.SAM.MetaInfo,1}:
 Bio.Align.SAM.MetaInfo:
    tag: SQ
  value: SN=Chr1 LN=30427671
 Bio.Align.SAM.MetaInfo:
    tag: SQ
  value: SN=Chr2 LN=19698289
 Bio.Align.SAM.MetaInfo:
    tag: SQ
  value: SN=Chr3 LN=23459830
 Bio.Align.SAM.MetaInfo:
    tag: SQ
  value: SN=Chr4 LN=18585056
 Bio.Align.SAM.MetaInfo:
    tag: SQ
  value: SN=Chr5 LN=26975502
 Bio.Align.SAM.MetaInfo:
    tag: SQ
  value: SN=chloroplast LN=154478
 Bio.Align.SAM.MetaInfo:
    tag: SQ
  value: SN=mitochondria LN=366924
</code></pre><p>In the above we can see there were 7 sequences in the reference: 5 chromosomes, one chloroplast sequence, and one mitochondrial sequence.</p><h2><a class="nav-anchor" id="SAM-and-BAM-Records-1" href="#SAM-and-BAM-Records-1">SAM and BAM Records</a></h2><p>BioAlignments supports the following accessors for <code>SAM.Record</code> types.</p><pre><code class="language-none">BioAlignments.SAM.flag
BioAlignments.SAM.ismapped
BioAlignments.SAM.isprimary
BioAlignments.SAM.refname
BioAlignments.SAM.position
BioAlignments.SAM.rightposition
BioAlignments.SAM.isnextmapped
BioAlignments.SAM.nextrefname
BioAlignments.SAM.nextposition
BioAlignments.SAM.mappingquality
BioAlignments.SAM.cigar
BioAlignments.SAM.alignment
BioAlignments.SAM.alignlength
BioAlignments.SAM.tempname
BioAlignments.SAM.templength
BioAlignments.SAM.sequence
BioAlignments.SAM.seqlength
BioAlignments.SAM.quality
BioAlignments.SAM.auxdata</code></pre><p>BioAlignments supports the following accessors for <code>BAM.Record</code> types.</p><pre><code class="language-none">BioAlignments.BAM.flag
BioAlignments.BAM.ismapped
BioAlignments.BAM.isprimary
BioAlignments.BAM.refid
BioAlignments.BAM.refname
BioAlignments.BAM.reflen
BioAlignments.BAM.position
BioAlignments.BAM.rightposition
BioAlignments.BAM.isnextmapped
BioAlignments.BAM.nextrefid
BioAlignments.BAM.nextrefname
BioAlignments.BAM.nextposition
BioAlignments.BAM.mappingquality
BioAlignments.BAM.cigar
BioAlignments.BAM.alignment
BioAlignments.BAM.alignlength
BioAlignments.BAM.tempname
BioAlignments.BAM.templength
BioAlignments.BAM.sequence
BioAlignments.BAM.seqlength
BioAlignments.BAM.quality
BioAlignments.BAM.auxdata</code></pre><h2><a class="nav-anchor" id="Accessing-auxiliary-data-1" href="#Accessing-auxiliary-data-1">Accessing auxiliary data</a></h2><p>SAM and BAM records support the storing of optional data fields associated with tags.</p><p>Tagged auxiliary data follows a format of <code>TAG:TYPE:VALUE</code>. <code>TAG</code> is a two-letter string, and each tag can only appear once per record. <code>TYPE</code> is a single case-sensetive letter which defined the format of <code>VALUE</code>.</p><table><tr><th>Type</th><th>Description</th></tr><tr><td>&#39;A&#39;</td><td>Printable character</td></tr><tr><td>&#39;i&#39;</td><td>Signed integer</td></tr><tr><td>&#39;f&#39;</td><td>Single-precision floating number</td></tr><tr><td>&#39;Z&#39;</td><td>Printable string, including space</td></tr><tr><td>&#39;H&#39;</td><td>Byte array in Hex format</td></tr><tr><td>&#39;B&#39;</td><td>Integer of numeric array</td></tr></table><p>For more information about these tags and their types we refer you to the [SAM/BAM specification][samtools-spec] and the additional [optional fields specification][samtags] document.</p><p>There are some tags that are reserved, predefined standard tags, for specific uses.</p><p>To access optional fields stored in tags, you use <code>getindex</code> indexing syntax on the record object. Note that accessing optional tag fields will result in type instability in Julia. This is because the type of the optional data is not known until run-time, as the tag is being read. This can have a significant impact on performance. To limit this, if the user knows the type of a value in advance, specifying it as a type annotation will alleviate the problem:</p><p>Below is an example of looping over records in a bam file and using indexing syntax to get the data stored in the &quot;NM&quot; tag. Note the <code>UInt8</code> type assertion to alleviate type instability.</p><pre><code class="language-julia">for record in open(BAM.Reader, &quot;data.bam&quot;)
    nm = record[&quot;NM&quot;]::UInt8
    # do something
end</code></pre><h2><a class="nav-anchor" id="Getting-records-in-a-range-1" href="#Getting-records-in-a-range-1">Getting records in a range</a></h2><p>BioAlignments supports the BAI index to fetch records in a specific range from a BAM file.  [Samtools][samtools] provides <code>index</code> subcommand to create an index file (.bai) from a sorted BAM file.</p><pre><code class="language-console">$ samtools index -b SRR1238088.sort.bam
$ ls SRR1238088.sort.bam*
SRR1238088.sort.bam     SRR1238088.sort.bam.bai</code></pre><p><code>eachoverlap(reader, chrom, range)</code> returns an iterator of BAM records overlapping the query interval:</p><pre><code class="language-julia">reader = open(BAM.Reader, &quot;SRR1238088.sort.bam&quot;, index=&quot;SRR1238088.sort.bam.bai&quot;)
for record in eachoverlap(reader, &quot;Chr2&quot;, 10000:11000)
    # `record` is a BAM.Record object
    # ...
end
close(reader)</code></pre><h2><a class="nav-anchor" id="Getting-records-overlapping-genomic-features-1" href="#Getting-records-overlapping-genomic-features-1">Getting records overlapping genomic features</a></h2><p><code>eachoverlap</code> also accepts the <code>Interval</code> type defined in <a href="https://github.com/BioJulia/GenomicFeatures.jl">GenomicFeatures.jl</a>.</p><p>This allows you to do things like first read in the genomic features from a GFF3 file, and then for each feature, iterate over all the BAM records that overlap with that feature.</p><pre><code class="language-julia"># Load GFF3 module.
using GenomicFeatures
using BioAlignments

# Load genomic features from a GFF3 file.
features = open(collect, GFF3.Reader, &quot;TAIR10_GFF3_genes.gff&quot;)

# Keep mRNA features.
filter!(x -&gt; GFF3.featuretype(x) == &quot;mRNA&quot;, features)

# Open a BAM file and iterate over records overlapping mRNA transcripts.
reader = open(BAM.Reader, &quot;SRR1238088.sort.bam&quot;, index = &quot;SRR1238088.sort.bam.bai&quot;)
for feature in features
    for record in eachoverlap(reader, feature)
        # `record` overlaps `feature`.
        # ...
    end
end
close(reader)</code></pre><h2><a class="nav-anchor" id="Writing-files-1" href="#Writing-files-1">Writing files</a></h2><p>In order to write a BAM or SAM file, you must first create a <code>SAM.Header</code>.</p><p>A <code>SAM.Header</code> is constructed from a vector of <code>SAM.MetaInfo</code> objects.</p><p>For example, to create the following simple header:</p><pre><code class="language-none">@HD VN:1.6 SO:coordinate
@SQ SN:ref LN:45</code></pre><pre><code class="language-julia">julia&gt; a = SAM.MetaInfo(&quot;HD&quot;, [&quot;VN&quot; =&gt; 1.6, &quot;SO&quot; =&gt; &quot;coordinate&quot;])
SAM.MetaInfo:
    tag: HD
  value: VN=1.6 SO=coordinate

julia&gt; b = SAM.MetaInfo(&quot;SQ&quot;, [&quot;SN&quot; =&gt; &quot;ref&quot;, &quot;LN&quot; =&gt; 45])
SAM.MetaInfo:
    tag: SQ
  value: SN=ref LN=45

julia&gt; h = SAM.Header([a, b])
SAM.Header(SAM.MetaInfo[SAM.MetaInfo:
    tag: HD
  value: VN=1.6 SO=coordinate, SAM.MetaInfo:
    tag: SQ
  value: SN=ref LN=45])
</code></pre><p>Then to create the writer for a SAM file, construct a <code>SAM.Writer</code> using the header and an <code>IO</code> type:</p><pre><code class="language-julia">julia&gt; samw = SAM.Writer(open(&quot;my-data.sam&quot;, &quot;w&quot;), h)
SAM.Writer(IOStream(&lt;file my-data.sam&gt;))
</code></pre><p>To make a BAM Writer is slightly different, as you need to use a specific stream type from the [BGZFStreams][bgzfstreams] package:</p><pre><code class="language-julia">julia&gt; using BGZFStreams

julia&gt; bamw = BAM.Writer(BGZFStream(open(&quot;my-data.bam&quot;, &quot;w&quot;), &quot;w&quot;))
BAM.Writer(BGZFStreams.BGZFStream{IOStream}(&lt;mode=write&gt;))
</code></pre><p>Once you have a BAM or SAM writer, you can use the <code>write</code> method to write <code>BAM.Record</code>s or <code>SAM.Record</code>s to file:</p><pre><code class="language-julia">julia&gt; write(bamw, rec) # Here rec is a `BAM.Record`
330780</code></pre><p>[samtools]:      https://samtools.github.io/ [samtools-spec]: https://samtools.github.io/hts-specs/SAMv1.pdf [samtags]:       https://samtools.github.io/hts-specs/SAMtags.pdf [bgzfstreams]:   https://github.com/BioJulia/BGZFStreams.jl</p><footer><hr/><a class="previous" href="pairalign.html"><span class="direction">Previous</span><span class="title">Pairwise alignment</span></a><a class="next" href="references.html"><span class="direction">Next</span><span class="title">API Reference</span></a></footer></article></body></html>
