<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>API Reference · BioAlignments.jl</title><link href="https://cdnjs.cloudflare.com/ajax/libs/normalize/4.2.0/normalize.min.css" rel="stylesheet" type="text/css"/><link href="https://fonts.googleapis.com/css?family=Lato|Roboto+Mono" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/styles/default.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL=".."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.2.0/require.min.js" data-main="../assets/documenter.js"></script><script src="../siteinfo.js"></script><script src="../../versions.js"></script><link href="../assets/documenter.css" rel="stylesheet" type="text/css"/></head><body><nav class="toc"><a href="../index.html"><img class="logo" src="../assets/logo.png" alt="BioAlignments.jl logo"/></a><h1>BioAlignments.jl</h1><select id="version-selector" onChange="window.location.href=this.value" style="visibility: hidden"></select><form class="search" id="search-form" action="../search/"><input id="search-query" name="q" type="text" placeholder="Search docs"/></form><ul><li><a class="toctext" href="../">Home</a></li><li><a class="toctext" href="../alignments/">Alignment representation</a></li><li><a class="toctext" href="../pairalign/">Pairwise alignment</a></li><li class="current"><a class="toctext" href>API Reference</a><ul class="internal"><li><a class="toctext" href="#Operations-1">Operations</a></li><li><a class="toctext" href="#Alignments-1">Alignments</a></li><li><a class="toctext" href="#Substitution-matrices-1">Substitution matrices</a></li><li><a class="toctext" href="#Pairwise-alignments-1">Pairwise alignments</a></li></ul></li></ul></nav><article id="docs"><header><nav><ul><li><a href>API Reference</a></li></ul><a class="edit-page" href="https://github.com/BioJulia/BioAlignments.jl/blob/develop/docs/src/references.md"><span class="fa"></span> Edit on GitHub</a></nav><hr/><div id="topbar"><span>API Reference</span><a class="fa fa-bars" href="#"></a></div></header><h1><a class="nav-anchor" id="API-Reference-1" href="#API-Reference-1">API Reference</a></h1><h2><a class="nav-anchor" id="Operations-1" href="#Operations-1">Operations</a></h2><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="BioAlignments.Operation" href="#BioAlignments.Operation"><code>BioAlignments.Operation</code></a> — <span class="docstring-category">Type</span>.</div><div><div><p>Alignment operation.</p></div></div><a class="source-link" target="_blank" href="https://github.com/BioJulia/BioAlignments.jl/blob/788cfe25bd08e99d0af490626c39255794fe8f19/src/operations.jl#L9-L11">source</a></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="BioAlignments.OP_MATCH" href="#BioAlignments.OP_MATCH"><code>BioAlignments.OP_MATCH</code></a> — <span class="docstring-category">Constant</span>.</div><div><div><p><code>&#39;M&#39;</code>: non-specific match</p></div></div><a class="source-link" target="_blank" href="https://github.com/BioJulia/BioAlignments.jl/blob/788cfe25bd08e99d0af490626c39255794fe8f19/src/operations.jl#L57">source</a></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="BioAlignments.OP_INSERT" href="#BioAlignments.OP_INSERT"><code>BioAlignments.OP_INSERT</code></a> — <span class="docstring-category">Constant</span>.</div><div><div><p><code>&#39;I&#39;</code>: insertion into reference sequence</p></div></div><a class="source-link" target="_blank" href="https://github.com/BioJulia/BioAlignments.jl/blob/788cfe25bd08e99d0af490626c39255794fe8f19/src/operations.jl#L57">source</a></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="BioAlignments.OP_DELETE" href="#BioAlignments.OP_DELETE"><code>BioAlignments.OP_DELETE</code></a> — <span class="docstring-category">Constant</span>.</div><div><div><p><code>&#39;D&#39;</code>: deletion from reference sequence</p></div></div><a class="source-link" target="_blank" href="https://github.com/BioJulia/BioAlignments.jl/blob/788cfe25bd08e99d0af490626c39255794fe8f19/src/operations.jl#L57">source</a></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="BioAlignments.OP_SKIP" href="#BioAlignments.OP_SKIP"><code>BioAlignments.OP_SKIP</code></a> — <span class="docstring-category">Constant</span>.</div><div><div><p><code>&#39;N&#39;</code>: (typically long) deletion from the reference, e.g. due to RNA splicing</p></div></div><a class="source-link" target="_blank" href="https://github.com/BioJulia/BioAlignments.jl/blob/788cfe25bd08e99d0af490626c39255794fe8f19/src/operations.jl#L57">source</a></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="BioAlignments.OP_SOFT_CLIP" href="#BioAlignments.OP_SOFT_CLIP"><code>BioAlignments.OP_SOFT_CLIP</code></a> — <span class="docstring-category">Constant</span>.</div><div><div><p><code>&#39;S&#39;</code>: sequence removed from the beginning or end of the query sequence but stored</p></div></div><a class="source-link" target="_blank" href="https://github.com/BioJulia/BioAlignments.jl/blob/788cfe25bd08e99d0af490626c39255794fe8f19/src/operations.jl#L57">source</a></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="BioAlignments.OP_HARD_CLIP" href="#BioAlignments.OP_HARD_CLIP"><code>BioAlignments.OP_HARD_CLIP</code></a> — <span class="docstring-category">Constant</span>.</div><div><div><p><code>&#39;H&#39;</code>: sequence removed from the beginning or end of the query sequence and not stored</p></div></div><a class="source-link" target="_blank" href="https://github.com/BioJulia/BioAlignments.jl/blob/788cfe25bd08e99d0af490626c39255794fe8f19/src/operations.jl#L57">source</a></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="BioAlignments.OP_PAD" href="#BioAlignments.OP_PAD"><code>BioAlignments.OP_PAD</code></a> — <span class="docstring-category">Constant</span>.</div><div><div><p><code>&#39;P&#39;</code>: not currently supported, but present for SAM/BAM compatibility</p></div></div><a class="source-link" target="_blank" href="https://github.com/BioJulia/BioAlignments.jl/blob/788cfe25bd08e99d0af490626c39255794fe8f19/src/operations.jl#L57">source</a></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="BioAlignments.OP_SEQ_MATCH" href="#BioAlignments.OP_SEQ_MATCH"><code>BioAlignments.OP_SEQ_MATCH</code></a> — <span class="docstring-category">Constant</span>.</div><div><div><p><code>&#39;=&#39;</code>: match operation with matching sequence positions</p></div></div><a class="source-link" target="_blank" href="https://github.com/BioJulia/BioAlignments.jl/blob/788cfe25bd08e99d0af490626c39255794fe8f19/src/operations.jl#L57">source</a></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="BioAlignments.OP_SEQ_MISMATCH" href="#BioAlignments.OP_SEQ_MISMATCH"><code>BioAlignments.OP_SEQ_MISMATCH</code></a> — <span class="docstring-category">Constant</span>.</div><div><div><p><code>&#39;X&#39;</code>: match operation with mismatching sequence positions</p></div></div><a class="source-link" target="_blank" href="https://github.com/BioJulia/BioAlignments.jl/blob/788cfe25bd08e99d0af490626c39255794fe8f19/src/operations.jl#L57">source</a></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="BioAlignments.OP_BACK" href="#BioAlignments.OP_BACK"><code>BioAlignments.OP_BACK</code></a> — <span class="docstring-category">Constant</span>.</div><div><div><p><code>&#39;B&#39;</code>: not currently supported, but present for SAM/BAM compatibility</p></div></div><a class="source-link" target="_blank" href="https://github.com/BioJulia/BioAlignments.jl/blob/788cfe25bd08e99d0af490626c39255794fe8f19/src/operations.jl#L57">source</a></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="BioAlignments.OP_START" href="#BioAlignments.OP_START"><code>BioAlignments.OP_START</code></a> — <span class="docstring-category">Constant</span>.</div><div><div><p><code>&#39;0&#39;</code>: indicate the start of an alignment within the reference and query sequence</p></div></div><a class="source-link" target="_blank" href="https://github.com/BioJulia/BioAlignments.jl/blob/788cfe25bd08e99d0af490626c39255794fe8f19/src/operations.jl#L57">source</a></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="BioAlignments.ismatchop" href="#BioAlignments.ismatchop"><code>BioAlignments.ismatchop</code></a> — <span class="docstring-category">Function</span>.</div><div><div><pre><code class="language-julia">ismatchop(op::Operation)</code></pre><p>Test if <code>op</code> is a match operation (i.e. <code>op ∈ (OP_MATCH, OP_SEQ_MATCH, OP_SEQ_MISMATCH)</code>).</p></div></div><a class="source-link" target="_blank" href="https://github.com/BioJulia/BioAlignments.jl/blob/788cfe25bd08e99d0af490626c39255794fe8f19/src/operations.jl#L70-L74">source</a></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="BioAlignments.isinsertop" href="#BioAlignments.isinsertop"><code>BioAlignments.isinsertop</code></a> — <span class="docstring-category">Function</span>.</div><div><div><pre><code class="language-julia">isinsertop(op::Operation)</code></pre><p>Test if <code>op</code> is a insertion operation (i.e. <code>op ∈ (OP_INSERT, OP_SOFT_CLIP, OP_HARD_CLIP)</code>).</p></div></div><a class="source-link" target="_blank" href="https://github.com/BioJulia/BioAlignments.jl/blob/788cfe25bd08e99d0af490626c39255794fe8f19/src/operations.jl#L79-L83">source</a></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="BioAlignments.isdeleteop" href="#BioAlignments.isdeleteop"><code>BioAlignments.isdeleteop</code></a> — <span class="docstring-category">Function</span>.</div><div><div><pre><code class="language-julia">isdeleteop(op::Operation)</code></pre><p>Test if <code>op</code> is a deletion operation (i.e. <code>op ∈ (OP_DELETE, OP_SKIP)</code>).</p></div></div><a class="source-link" target="_blank" href="https://github.com/BioJulia/BioAlignments.jl/blob/788cfe25bd08e99d0af490626c39255794fe8f19/src/operations.jl#L88-L92">source</a></section><h2><a class="nav-anchor" id="Alignments-1" href="#Alignments-1">Alignments</a></h2><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="BioAlignments.AlignmentAnchor" href="#BioAlignments.AlignmentAnchor"><code>BioAlignments.AlignmentAnchor</code></a> — <span class="docstring-category">Type</span>.</div><div><div><p>Alignment operation with anchoring positions.</p></div></div><a class="source-link" target="_blank" href="https://github.com/BioJulia/BioAlignments.jl/blob/788cfe25bd08e99d0af490626c39255794fe8f19/src/anchors.jl#L9-L11">source</a></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="BioAlignments.Alignment" href="#BioAlignments.Alignment"><code>BioAlignments.Alignment</code></a> — <span class="docstring-category">Type</span>.</div><div><div><p>Alignment of two sequences.</p></div></div><a class="source-link" target="_blank" href="https://github.com/BioJulia/BioAlignments.jl/blob/788cfe25bd08e99d0af490626c39255794fe8f19/src/alignment.jl#L9-L11">source</a></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="BioAlignments.Alignment-Tuple{Array{AlignmentAnchor,1},Bool}" href="#BioAlignments.Alignment-Tuple{Array{AlignmentAnchor,1},Bool}"><code>BioAlignments.Alignment</code></a> — <span class="docstring-category">Method</span>.</div><div><div><pre><code class="language-julia">Alignment(anchors::Vector{AlignmentAnchor}, check=true)</code></pre><p>Create an alignment object from a sequence of alignment anchors.</p></div></div><a class="source-link" target="_blank" href="https://github.com/BioJulia/BioAlignments.jl/blob/788cfe25bd08e99d0af490626c39255794fe8f19/src/alignment.jl#L18-L22">source</a></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="BioAlignments.Alignment-Tuple{AbstractString,Int64,Int64}" href="#BioAlignments.Alignment-Tuple{AbstractString,Int64,Int64}"><code>BioAlignments.Alignment</code></a> — <span class="docstring-category">Method</span>.</div><div><div><pre><code class="language-julia">Alignment(cigar::AbstractString, seqpos=1, refpos=1)</code></pre><p>Make an alignment object from a CIGAR string.</p><p><code>seqpos</code> and <code>refpos</code> specify the starting positions of two sequences.</p></div></div><a class="source-link" target="_blank" href="https://github.com/BioJulia/BioAlignments.jl/blob/788cfe25bd08e99d0af490626c39255794fe8f19/src/alignment.jl#L50-L56">source</a></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="BioAlignments.seq2ref-Tuple{Alignment,Integer}" href="#BioAlignments.seq2ref-Tuple{Alignment,Integer}"><code>BioAlignments.seq2ref</code></a> — <span class="docstring-category">Method</span>.</div><div><div><pre><code class="language-julia">seq2ref(aln::Alignment, i::Integer)::Tuple{Int,Operation}</code></pre><p>Map a position <code>i</code> from sequence to reference.</p></div></div><a class="source-link" target="_blank" href="https://github.com/BioJulia/BioAlignments.jl/blob/788cfe25bd08e99d0af490626c39255794fe8f19/src/alignment.jl#L103-L107">source</a></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="BioAlignments.ref2seq-Tuple{Alignment,Integer}" href="#BioAlignments.ref2seq-Tuple{Alignment,Integer}"><code>BioAlignments.ref2seq</code></a> — <span class="docstring-category">Method</span>.</div><div><div><pre><code class="language-julia">ref2seq(aln::Alignment, i::Integer)::Tuple{Int,Operation}</code></pre><p>Map a position <code>i</code> from reference to sequence.</p></div></div><a class="source-link" target="_blank" href="https://github.com/BioJulia/BioAlignments.jl/blob/788cfe25bd08e99d0af490626c39255794fe8f19/src/alignment.jl#L121-L125">source</a></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="BioAlignments.cigar-Tuple{Alignment}" href="#BioAlignments.cigar-Tuple{Alignment}"><code>BioAlignments.cigar</code></a> — <span class="docstring-category">Method</span>.</div><div><div><pre><code class="language-julia">cigar(aln::Alignment)</code></pre><p>Make a CIGAR string encoding of <code>aln</code>.</p><p>This is not entirely lossless as it discards the alignments start positions.</p></div></div><a class="source-link" target="_blank" href="https://github.com/BioJulia/BioAlignments.jl/blob/788cfe25bd08e99d0af490626c39255794fe8f19/src/alignment.jl#L139-L145">source</a></section><h2><a class="nav-anchor" id="Substitution-matrices-1" href="#Substitution-matrices-1">Substitution matrices</a></h2><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="BioAlignments.AbstractSubstitutionMatrix" href="#BioAlignments.AbstractSubstitutionMatrix"><code>BioAlignments.AbstractSubstitutionMatrix</code></a> — <span class="docstring-category">Type</span>.</div><div><div><p>Supertype of substitution matrix.</p><p>The required method:</p><ul><li><code>Base.getindex(submat, x, y)</code>: substitution score/cost from <code>x</code> to <code>y</code></li></ul></div></div><a class="source-link" target="_blank" href="https://github.com/BioJulia/BioAlignments.jl/blob/788cfe25bd08e99d0af490626c39255794fe8f19/src/submat.jl#L9-L15">source</a></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="BioAlignments.SubstitutionMatrix" href="#BioAlignments.SubstitutionMatrix"><code>BioAlignments.SubstitutionMatrix</code></a> — <span class="docstring-category">Type</span>.</div><div><div><p>Substitution matrix.</p></div></div><a class="source-link" target="_blank" href="https://github.com/BioJulia/BioAlignments.jl/blob/788cfe25bd08e99d0af490626c39255794fe8f19/src/submat.jl#L18-L20">source</a></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="BioAlignments.DichotomousSubstitutionMatrix" href="#BioAlignments.DichotomousSubstitutionMatrix"><code>BioAlignments.DichotomousSubstitutionMatrix</code></a> — <span class="docstring-category">Type</span>.</div><div><div><p>Dichotomous substitution matrix.</p></div></div><a class="source-link" target="_blank" href="https://github.com/BioJulia/BioAlignments.jl/blob/788cfe25bd08e99d0af490626c39255794fe8f19/src/submat.jl#L149-L151">source</a></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="BioAlignments.EDNAFULL" href="#BioAlignments.EDNAFULL"><code>BioAlignments.EDNAFULL</code></a> — <span class="docstring-category">Constant</span>.</div><div><div><p>EDNAFULL (or NUC4.4) substitution matrix</p></div></div><a class="source-link" target="_blank" href="https://github.com/BioJulia/BioAlignments.jl/blob/788cfe25bd08e99d0af490626c39255794fe8f19/src/submat.jl#L220">source</a></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="BioAlignments.PAM30" href="#BioAlignments.PAM30"><code>BioAlignments.PAM30</code></a> — <span class="docstring-category">Constant</span>.</div><div><div><p>PAM30 substitution matrix</p></div></div><a class="source-link" target="_blank" href="https://github.com/BioJulia/BioAlignments.jl/blob/788cfe25bd08e99d0af490626c39255794fe8f19/src/submat.jl#L223">source</a></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="BioAlignments.PAM70" href="#BioAlignments.PAM70"><code>BioAlignments.PAM70</code></a> — <span class="docstring-category">Constant</span>.</div><div><div><p>PAM70 substitution matrix</p></div></div><a class="source-link" target="_blank" href="https://github.com/BioJulia/BioAlignments.jl/blob/788cfe25bd08e99d0af490626c39255794fe8f19/src/submat.jl#L226">source</a></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="BioAlignments.PAM250" href="#BioAlignments.PAM250"><code>BioAlignments.PAM250</code></a> — <span class="docstring-category">Constant</span>.</div><div><div><p>PAM250 substitution matrix</p></div></div><a class="source-link" target="_blank" href="https://github.com/BioJulia/BioAlignments.jl/blob/788cfe25bd08e99d0af490626c39255794fe8f19/src/submat.jl#L229">source</a></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="BioAlignments.BLOSUM45" href="#BioAlignments.BLOSUM45"><code>BioAlignments.BLOSUM45</code></a> — <span class="docstring-category">Constant</span>.</div><div><div><p>BLOSUM45 substitution matrix</p></div></div><a class="source-link" target="_blank" href="https://github.com/BioJulia/BioAlignments.jl/blob/788cfe25bd08e99d0af490626c39255794fe8f19/src/submat.jl#L232">source</a></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="BioAlignments.BLOSUM50" href="#BioAlignments.BLOSUM50"><code>BioAlignments.BLOSUM50</code></a> — <span class="docstring-category">Constant</span>.</div><div><div><p>BLOSUM50 substitution matrix</p></div></div><a class="source-link" target="_blank" href="https://github.com/BioJulia/BioAlignments.jl/blob/788cfe25bd08e99d0af490626c39255794fe8f19/src/submat.jl#L235">source</a></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="BioAlignments.BLOSUM62" href="#BioAlignments.BLOSUM62"><code>BioAlignments.BLOSUM62</code></a> — <span class="docstring-category">Constant</span>.</div><div><div><p>BLOSUM62 substitution matrix</p></div></div><a class="source-link" target="_blank" href="https://github.com/BioJulia/BioAlignments.jl/blob/788cfe25bd08e99d0af490626c39255794fe8f19/src/submat.jl#L238">source</a></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="BioAlignments.BLOSUM80" href="#BioAlignments.BLOSUM80"><code>BioAlignments.BLOSUM80</code></a> — <span class="docstring-category">Constant</span>.</div><div><div><p>BLOSUM80 substitution matrix</p></div></div><a class="source-link" target="_blank" href="https://github.com/BioJulia/BioAlignments.jl/blob/788cfe25bd08e99d0af490626c39255794fe8f19/src/submat.jl#L241">source</a></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="BioAlignments.BLOSUM90" href="#BioAlignments.BLOSUM90"><code>BioAlignments.BLOSUM90</code></a> — <span class="docstring-category">Constant</span>.</div><div><div><p>BLOSUM90 substitution matrix</p></div></div><a class="source-link" target="_blank" href="https://github.com/BioJulia/BioAlignments.jl/blob/788cfe25bd08e99d0af490626c39255794fe8f19/src/submat.jl#L244">source</a></section><h2><a class="nav-anchor" id="Pairwise-alignments-1" href="#Pairwise-alignments-1">Pairwise alignments</a></h2><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="BioAlignments.PairwiseAlignment" href="#BioAlignments.PairwiseAlignment"><code>BioAlignments.PairwiseAlignment</code></a> — <span class="docstring-category">Type</span>.</div><div><div><p>Pairwise alignment</p></div></div><a class="source-link" target="_blank" href="https://github.com/BioJulia/BioAlignments.jl/blob/788cfe25bd08e99d0af490626c39255794fe8f19/src/pairwise/alignment.jl#L9-L11">source</a></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="Base.count-Tuple{PairwiseAlignment,Operation}" href="#Base.count-Tuple{PairwiseAlignment,Operation}"><code>Base.count</code></a> — <span class="docstring-category">Method</span>.</div><div><div><pre><code class="language-julia">count(aln::PairwiseAlignment, target::Operation)</code></pre><p>Count the number of positions where the <code>target</code> operation is applied.</p></div></div><a class="source-link" target="_blank" href="https://github.com/BioJulia/BioAlignments.jl/blob/788cfe25bd08e99d0af490626c39255794fe8f19/src/pairwise/alignment.jl#L65-L69">source</a></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="BioAlignments.count_matches" href="#BioAlignments.count_matches"><code>BioAlignments.count_matches</code></a> — <span class="docstring-category">Function</span>.</div><div><div><pre><code class="language-julia">count_matches(aln)</code></pre><p>Count the number of matching positions.</p></div></div><a class="source-link" target="_blank" href="https://github.com/BioJulia/BioAlignments.jl/blob/788cfe25bd08e99d0af490626c39255794fe8f19/src/pairwise/alignment.jl#L86-L90">source</a></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="BioAlignments.count_mismatches" href="#BioAlignments.count_mismatches"><code>BioAlignments.count_mismatches</code></a> — <span class="docstring-category">Function</span>.</div><div><div><pre><code class="language-julia">count_mismatches(aln)</code></pre><p>Count the number of mismatching positions.</p></div></div><a class="source-link" target="_blank" href="https://github.com/BioJulia/BioAlignments.jl/blob/788cfe25bd08e99d0af490626c39255794fe8f19/src/pairwise/alignment.jl#L93-L97">source</a></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="BioAlignments.count_insertions" href="#BioAlignments.count_insertions"><code>BioAlignments.count_insertions</code></a> — <span class="docstring-category">Function</span>.</div><div><div><pre><code class="language-julia">count_insertions(aln)</code></pre><p>Count the number of inserting positions.</p></div></div><a class="source-link" target="_blank" href="https://github.com/BioJulia/BioAlignments.jl/blob/788cfe25bd08e99d0af490626c39255794fe8f19/src/pairwise/alignment.jl#L100-L104">source</a></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="BioAlignments.count_deletions" href="#BioAlignments.count_deletions"><code>BioAlignments.count_deletions</code></a> — <span class="docstring-category">Function</span>.</div><div><div><pre><code class="language-julia">count_deletions(aln)</code></pre><p>Count the number of deleting positions.</p></div></div><a class="source-link" target="_blank" href="https://github.com/BioJulia/BioAlignments.jl/blob/788cfe25bd08e99d0af490626c39255794fe8f19/src/pairwise/alignment.jl#L107-L111">source</a></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="BioAlignments.count_aligned" href="#BioAlignments.count_aligned"><code>BioAlignments.count_aligned</code></a> — <span class="docstring-category">Function</span>.</div><div><div><pre><code class="language-julia">count_aligned(aln)</code></pre><p>Count the number of aligned positions.</p></div></div><a class="source-link" target="_blank" href="https://github.com/BioJulia/BioAlignments.jl/blob/788cfe25bd08e99d0af490626c39255794fe8f19/src/pairwise/alignment.jl#L114-L118">source</a></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="BioAlignments.GlobalAlignment" href="#BioAlignments.GlobalAlignment"><code>BioAlignments.GlobalAlignment</code></a> — <span class="docstring-category">Type</span>.</div><div><div><p>Global-global alignment with end gap penalties.</p></div></div><a class="source-link" target="_blank" href="https://github.com/BioJulia/BioAlignments.jl/blob/788cfe25bd08e99d0af490626c39255794fe8f19/src/types.jl#L15">source</a></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="BioAlignments.SemiGlobalAlignment" href="#BioAlignments.SemiGlobalAlignment"><code>BioAlignments.SemiGlobalAlignment</code></a> — <span class="docstring-category">Type</span>.</div><div><div><p>Global-local alignment.</p></div></div><a class="source-link" target="_blank" href="https://github.com/BioJulia/BioAlignments.jl/blob/788cfe25bd08e99d0af490626c39255794fe8f19/src/types.jl#L18">source</a></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="BioAlignments.OverlapAlignment" href="#BioAlignments.OverlapAlignment"><code>BioAlignments.OverlapAlignment</code></a> — <span class="docstring-category">Type</span>.</div><div><div><p>Global-global alignment without end gap penalties.</p></div></div><a class="source-link" target="_blank" href="https://github.com/BioJulia/BioAlignments.jl/blob/788cfe25bd08e99d0af490626c39255794fe8f19/src/types.jl#L21">source</a></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="BioAlignments.LocalAlignment" href="#BioAlignments.LocalAlignment"><code>BioAlignments.LocalAlignment</code></a> — <span class="docstring-category">Type</span>.</div><div><div><p>Local-local alignment.</p></div></div><a class="source-link" target="_blank" href="https://github.com/BioJulia/BioAlignments.jl/blob/788cfe25bd08e99d0af490626c39255794fe8f19/src/types.jl#L24">source</a></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="BioAlignments.EditDistance" href="#BioAlignments.EditDistance"><code>BioAlignments.EditDistance</code></a> — <span class="docstring-category">Type</span>.</div><div><div><p>Edit distance.</p></div></div><a class="source-link" target="_blank" href="https://github.com/BioJulia/BioAlignments.jl/blob/788cfe25bd08e99d0af490626c39255794fe8f19/src/types.jl#L31">source</a></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="BioAlignments.HammingDistance" href="#BioAlignments.HammingDistance"><code>BioAlignments.HammingDistance</code></a> — <span class="docstring-category">Type</span>.</div><div><div><p>Hamming distance.</p><p>A special case of <code>EditDistance</code> with the costs of insertion and deletion are infinitely large.</p></div></div><a class="source-link" target="_blank" href="https://github.com/BioJulia/BioAlignments.jl/blob/788cfe25bd08e99d0af490626c39255794fe8f19/src/types.jl#L42-L47">source</a></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="BioAlignments.LevenshteinDistance" href="#BioAlignments.LevenshteinDistance"><code>BioAlignments.LevenshteinDistance</code></a> — <span class="docstring-category">Type</span>.</div><div><div><p>Levenshtein distance.</p><p>A special case of <code>EditDistance</code> with the costs of mismatch, insertion, and deletion are 1.</p></div></div><a class="source-link" target="_blank" href="https://github.com/BioJulia/BioAlignments.jl/blob/788cfe25bd08e99d0af490626c39255794fe8f19/src/types.jl#L34-L39">source</a></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="BioAlignments.AbstractScoreModel" href="#BioAlignments.AbstractScoreModel"><code>BioAlignments.AbstractScoreModel</code></a> — <span class="docstring-category">Type</span>.</div><div><div><p>Supertype of score model.</p></div></div><a class="source-link" target="_blank" href="https://github.com/BioJulia/BioAlignments.jl/blob/788cfe25bd08e99d0af490626c39255794fe8f19/src/models.jl#L13-L15">source</a></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="BioAlignments.AffineGapScoreModel" href="#BioAlignments.AffineGapScoreModel"><code>BioAlignments.AffineGapScoreModel</code></a> — <span class="docstring-category">Type</span>.</div><div><div><pre><code class="language-julia">AffineGapScoreModel(submat, gap_open, gap_extend)
AffineGapScoreModel(submat, gap_open=, gap_extend=)
AffineGapScoreModel(match=, mismatch=, gap_open=, gap_extend=)</code></pre><p>Affine gap scoring model.</p><p>This creates an affine gap scroing model object for alignment from a substitution matrix (<code>submat</code>), a gap opening score (<code>gap_open</code>), and a gap extending score (<code>gap_extend</code>). A consecutive gap of length <code>k</code> has a score of <code>gap_open + gap_extend * k</code>. Note that both of the gap scores should be non-positive.  As a shorthand of creating a dichotomous substitution matrix, you can write as, for example, <code>AffineGapScoreModel(match=5, mismatch=-3, gap_open=-2, gap_extend=-1)</code>.</p><p><strong>Example</strong></p><pre><code class="language-none">using BioSequences
using BioAlignments

# create an affine gap scoring model from a predefined substitution
# matrix and gap opening/extending scores.
affinegap = AffineGapScoreModel(BLOSUM62, gap_open=-10, gap_extend=-1)

# run global alignment between two amino acid sequenecs
pairalign(GlobalAlignment(), aa&quot;IDGAAGQQL&quot;, aa&quot;IDGATGQL&quot;, affinegap)</code></pre><p>See also: <code>SubstitutionMatrix</code>, <code>pairalign</code>, <code>CostModel</code></p></div></div><a class="source-link" target="_blank" href="https://github.com/BioJulia/BioAlignments.jl/blob/788cfe25bd08e99d0af490626c39255794fe8f19/src/models.jl#L18-L47">source</a></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="BioAlignments.AbstractCostModel" href="#BioAlignments.AbstractCostModel"><code>BioAlignments.AbstractCostModel</code></a> — <span class="docstring-category">Type</span>.</div><div><div><p>Supertype of cost model.</p></div></div><a class="source-link" target="_blank" href="https://github.com/BioJulia/BioAlignments.jl/blob/788cfe25bd08e99d0af490626c39255794fe8f19/src/models.jl#L123-L125">source</a></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="BioAlignments.CostModel" href="#BioAlignments.CostModel"><code>BioAlignments.CostModel</code></a> — <span class="docstring-category">Type</span>.</div><div><div><pre><code class="language-julia">CostModel(submat, insertion, deletion)
CostModel(submat, insertion=, deletion=)
CostModel(match=, mismatch=, insertion=, deletion=)</code></pre><p>Cost model.</p><p>This creates a cost model object for alignment from substitution matrix (<code>submat</code>), an insertion cost (<code>insertion</code>), and a deletion cost (<code>deletion</code>). Note that both of the insertion and deletion costs should be non-negative.  As a shorthand of creating a dichotomous substitution matrix, you can write as, for example, <code>CostModel(match=0, mismatch=1, insertion=2, deletion=2)</code>.</p><p><strong>Example</strong></p><pre><code class="language-none">using BioAlignments

# create a cost model from a substitution matrix and indel costs
cost = CostModel(ones(128, 128) - eye(128), insertion=.5, deletion=.5)

# run global alignment to minimize edit distance
pairalign(EditDistance(), &quot;intension&quot;, &quot;execution&quot;, cost)</code></pre><p>See also: <code>SubstitutionMatrix</code>, <code>pairalign</code>, <code>AffineGapScoreModel</code></p></div></div><a class="source-link" target="_blank" href="https://github.com/BioJulia/BioAlignments.jl/blob/788cfe25bd08e99d0af490626c39255794fe8f19/src/models.jl#L128-L153">source</a></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="BioAlignments.PairwiseAlignmentResult" href="#BioAlignments.PairwiseAlignmentResult"><code>BioAlignments.PairwiseAlignmentResult</code></a> — <span class="docstring-category">Type</span>.</div><div><div><p>Result of pairwise alignment</p></div></div><a class="source-link" target="_blank" href="https://github.com/BioJulia/BioAlignments.jl/blob/788cfe25bd08e99d0af490626c39255794fe8f19/src/pairwise/result.jl#L9-L11">source</a></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="BioAlignments.pairalign" href="#BioAlignments.pairalign"><code>BioAlignments.pairalign</code></a> — <span class="docstring-category">Function</span>.</div><div><div><pre><code class="language-julia">pairalign(type, seq, ref, model, [options...])</code></pre><p>Run pairwise alignment between two sequences: <code>seq</code> and <code>ref</code>.</p><p>Available <code>type</code>s are:</p><ul><li><code>GlobalAlignment()</code></li><li><code>LocalAlignment()</code></li><li><code>SemiGlobalAlignment()</code></li><li><code>OverlapAlignment()</code></li><li><code>EditDistance()</code></li><li><code>LevenshteinDistance()</code></li><li><code>HammingDistance()</code></li></ul><p><code>GlobalAlignment</code>, <code>LocalAlignment</code>, <code>SemiGlobalAlignment</code>, and <code>OverlapAlignment</code> are problem that maximizes alignment score between two sequences.  Therefore, <code>model</code> should be an instance of <code>AbstractScoreModel</code> (e.g. <code>AffineGapScoreModel</code>).</p><p><code>EditDistance</code>, <code>LevenshteinDistance</code>, and <code>HammingDistance</code> are problem that minimizes alignment cost between two sequences.  As for <code>EditDistance</code>, <code>model</code> should be an instance of <code>AbstractCostModel</code> (e.g. <code>CostModel</code>). <code>LevenshteinDistance</code> and <code>HammingDistance</code> have predefined a cost model, so users cannot specify a cost model for these alignment types.</p><p>When you pass the <code>score_only=true</code> or <code>distance_only=true</code> option to <code>pairalign</code>, the result of pairwise alignment holds alignment score/distance only.  This may enable some algorithms to run faster than calculating full alignment result.  Other available <code>options</code> are documented for each alignemnt type.</p><p><strong>Example</strong></p><pre><code class="language-none">using BioSequences
using BioAlignments

# create affine gap scoring model
affinegap = AffineGapScoreModel(
    match=5,
    mismatch=-4,
    gap_open=-5,
    gap_extend=-3
)

# run global alignment between two DNA sequences
pairalign(GlobalAlignment(), dna&quot;AGGTAG&quot;, dna&quot;ATTG&quot;, affinegap)

# run local alignment between two DNA sequences
pairalign(LocalAlignment(), dna&quot;AGGTAG&quot;, dna&quot;ATTG&quot;, affinegap)

# you cannot specify a cost model in LevenshteinDistance
pairalign(LevenshteinDistance(), dna&quot;AGGTAG&quot;, dna&quot;ATTG&quot;)</code></pre><p>See also: <code>AffineGapScoreModel</code>, <code>CostModel</code></p></div></div><a class="source-link" target="_blank" href="https://github.com/BioJulia/BioAlignments.jl/blob/788cfe25bd08e99d0af490626c39255794fe8f19/src/pairwise/pairalign.jl#L19-L75">source</a></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="BioAlignments.score" href="#BioAlignments.score"><code>BioAlignments.score</code></a> — <span class="docstring-category">Function</span>.</div><div><div><pre><code class="language-julia">score(alignment_result)</code></pre><p>Return score of alignment.</p></div></div><a class="source-link" target="_blank" href="https://github.com/BioJulia/BioAlignments.jl/blob/788cfe25bd08e99d0af490626c39255794fe8f19/src/pairwise/result.jl#L31-L35">source</a></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="BioGenerics.distance" href="#BioGenerics.distance"><code>BioGenerics.distance</code></a> — <span class="docstring-category">Function</span>.</div><div><div><pre><code class="language-julia">distance(alignment_result)</code></pre><p>Retrun distance of alignment.</p></div></div><a class="source-link" target="_blank" href="https://github.com/BioJulia/BioAlignments.jl/blob/788cfe25bd08e99d0af490626c39255794fe8f19/src/pairwise/result.jl#L38-L42">source</a></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="BioAlignments.alignment" href="#BioAlignments.alignment"><code>BioAlignments.alignment</code></a> — <span class="docstring-category">Function</span>.</div><div><div><pre><code class="language-julia">alignment(alignment_result)</code></pre><p>Return alignment if any.</p><p>See also: <code>hasalignment</code></p></div></div><a class="source-link" target="_blank" href="https://github.com/BioJulia/BioAlignments.jl/blob/788cfe25bd08e99d0af490626c39255794fe8f19/src/pairwise/result.jl#L53-L59">source</a></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="BioAlignments.hasalignment" href="#BioAlignments.hasalignment"><code>BioAlignments.hasalignment</code></a> — <span class="docstring-category">Function</span>.</div><div><div><pre><code class="language-julia">hasalignment(alignment_result)</code></pre><p>Check if alignment is stored or not.</p></div></div><a class="source-link" target="_blank" href="https://github.com/BioJulia/BioAlignments.jl/blob/788cfe25bd08e99d0af490626c39255794fe8f19/src/pairwise/result.jl#L46-L50">source</a></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="BioAlignments.seq2ref-Tuple{PairwiseAlignment,Integer}" href="#BioAlignments.seq2ref-Tuple{PairwiseAlignment,Integer}"><code>BioAlignments.seq2ref</code></a> — <span class="docstring-category">Method</span>.</div><div><div><pre><code class="language-julia">seq2ref(aln::PairwiseAlignment, i::Integer)::Tuple{Int,Operation}</code></pre><p>Map a position <code>i</code> from the first sequence to the second.</p></div></div><a class="source-link" target="_blank" href="https://github.com/BioJulia/BioAlignments.jl/blob/788cfe25bd08e99d0af490626c39255794fe8f19/src/pairwise/alignment.jl#L133-L137">source</a></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="BioAlignments.ref2seq-Tuple{PairwiseAlignment,Integer}" href="#BioAlignments.ref2seq-Tuple{PairwiseAlignment,Integer}"><code>BioAlignments.ref2seq</code></a> — <span class="docstring-category">Method</span>.</div><div><div><pre><code class="language-julia">ref2seq(aln::PairwiseAlignment, i::Integer)::Tuple{Int,Operation}</code></pre><p>Map a position <code>i</code> from the second sequence to the first.</p></div></div><a class="source-link" target="_blank" href="https://github.com/BioJulia/BioAlignments.jl/blob/788cfe25bd08e99d0af490626c39255794fe8f19/src/pairwise/alignment.jl#L142-L146">source</a></section><footer><hr/><a class="previous" href="../pairalign/"><span class="direction">Previous</span><span class="title">Pairwise alignment</span></a></footer></article></body></html>
